packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
   }
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+  * relayer? TODO determine if we can predict this request (question for komenci folks)
+  * gatewayFee / gateWayFeeRecipient? (TODO: ask Contracts if better to omit these or add them for forward compatibility) 
+  */
+
+  /**
+   * @notice Executes a refundable meta-transaction on behalf of the signer.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param maxGasPrice The maximum gas price the user is willing to pay.
+   * @param gasLimit The gas limit for entire relayed transaction.
+   * @param metaGasLimit The gas limit for just the meta-transaction.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The return value of the meta-transaction execution.
+   */
+  function executeMetaTransactionWithRefund(
+    address destination,
+    uint256 value,
+    bytes calldata data,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit == gasLeft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
+    require(msg.sender == tx.origin, "relayer must be EOA");
+    require(gasLimit <= block.gaslimit, "Impossible gas limit");
+    require(metaGasLimit < gasLimit, "metaGasLimit must be less than gasLimit");
+    if (data.length > 0) require(Address.isContract(destination), "Invalid contract address");
+
+    // checking gasCurrency is out of scope, would require pre-compile
```

Comment:
Haven't thought through this all the way, but I'm wondering if we maybe could implement specifying a gas currency without the need for an on-chain check here? As long as the MTW owner specifies the currency they will refund with, the gasLimit and gasPrice, they won't get charged more than they expected, so it doesn't really matter what currency the relayer actually used.

Reply:
+1 to this. The parameters set by the user can be seen as an agreement from the user to pay the relayer at some rate per unit of gas used, and this could be in any token, even those that are not whitelisted. If the relayer then wants to pay for the transaction gas fees in a different currency, that should be perfectly fine.

Reply:
(In this case we may not be able to easily compare the gas price on the transaction overall, and that of the refund because they may be different currencies. Two options to address this would be to either use the maxGasPrice specified by the user as the refund rate, or add a relayer specified refund gas price parameter that can be set up to the max gas price. (Downside to both of these is that the relayer can arbitrage any difference between the user-specified max gas price and the one they cleared the transaction at when they are in different currencies). Another option would be to actually check the on-chain oracle for the exchange rate to compare gas prices.

Reply:
If I can try to summarize:
- No matter what, the signer should specify the refund currency, and (max)gasPrice/gasLimit denominated in that currency. 
- If the relayer wants to pay gas in a different currency, that's okay as long as we can ensure they are refunded as specified by the signer. 
- If the relayer chooses to pay in a different currency, then tx.gasprice (specified by the relayer) will not be denominated in the same currency as the signer specified (max)gasPrice. 
- If the relayer uses a gas currency that is more valuable on a unit basis, that is fine since tx.gasprice < (max)gasPrice. They can do this with the implementation we have already. 
- If we want to allow the relayer to pay for gas in a currency that is less valuable on a unit basis, then we will need to use an oracle to convert the gas prices into the same units for comparison.
- In order to use an oracle, we need to know what currency the relayer is paying for gas in because otherwise they could claim to use a less valuable currency than they actually submit the transaction with, and get refunded at a higher rate than they should be. 
- We cannot know what gas currency the relayer actually submits the transaction with because it is not exposed to the smart contracts.
- Unless I've made an error above (I may well have) I think this means we can't support this feature without exposing tx.gascurrency. 

Does this reasoning check out?